FreeRTOS_IP.c

```diff
@@ -4204,6 +4287,471 @@ const char * FreeRTOS_strerror_r( BaseType_t xErrnum,
+           ( ( ( uint32_t ) pucPtr[ 2 ] ) << 8 ) |
+           ( ( ( uint32_t ) pucPtr[ 3 ] ) );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Convert character array (of size 2) to equivalent 16-bit value.
+ * @param[in] pucPtr: The character array.
+ * @return 16-bit equivalent value extracted from the character array.
+ *
+ * @note Going by MISRA rules, these utility functions should not be defined
+ *        if they are not being used anywhere. But their use depends on the
+ *        application and hence these functions are defined unconditionally.
+ */
+uint16_t usChar2u16( const uint8_t * pucPtr )
+{
+    return ( uint16_t )
+           ( ( ( ( uint32_t ) pucPtr[ 0 ] ) << 8 ) |
+             ( ( ( uint32_t ) pucPtr[ 1 ] ) ) );
+}
+
+#if ( ipconfigUSE_IPv6 != 0 )
+
+/**
+ * @brief Get the size of the IP-header, by checking the type of the network buffer.
+ * @param[in] pxNetworkBuffer: The network buffer.
+ * @return The size of the corresponding IP-header.
+ */
+    size_t uxIPHeaderSizePacket( const NetworkBufferDescriptor_t * pxNetworkBuffer )
+    {
+        size_t uxResult;
+        /* Map the buffer onto Ethernet Header struct for easy access to fields. */
+        /* misra_c_2012_rule_11_3_violation */
+        const EthernetHeader_t * pxHeader = ipCAST_CONST_PTR_TO_CONST_TYPE_PTR( EthernetHeader_t, pxNetworkBuffer->pucEthernetBuffer );
+
+        if( pxHeader->usFrameType == ( uint16_t ) ipIPv6_FRAME_TYPE )
+        {
+            uxResult = ipSIZE_OF_IPv6_HEADER;
+        }
+        else
+        {
+            uxResult = ipSIZE_OF_IPv4_HEADER;
+        }
+
+        return uxResult;
+    }
+#else /* if ( ipconfigUSE_IPv6 != 0 ) */
+
+/**
+ * @brief Get the size of an IPv4-header, independent of the network buffer.
+ * @param[in] pxNetworkBuffer: The network buffer ( ignored ).
+ * @return The size of the corresponding IP-header.
+ */
+    size_t uxIPHeaderSizePacket( const NetworkBufferDescriptor_t * pxNetworkBuffer )
+    {
+        return ipSIZE_OF_IPv4_HEADER;
+    }
+#endif /* if ( ipconfigUSE_IPv6 != 0 ) */
+
+#if ( ipconfigUSE_IPv6 != 0 )
+
+/**
+ * @brief Get the size of the IP-header, by checking if the socket bIsIPv6 set.
+ * @param[in] pxSocket: The socket.
+ * @return The size of the corresponding IP-header.
+ */
+    size_t uxIPHeaderSizeSocket( const FreeRTOS_Socket_t * pxSocket )
+    {
+        size_t uxResult;
+
+        if( ( pxSocket != NULL ) && ( pxSocket->bits.bIsIPv6 != pdFALSE_UNSIGNED ) )
+        {
+            uxResult = ipSIZE_OF_IPv6_HEADER;
+        }
+        else
+        {
+            uxResult = ipSIZE_OF_IPv4_HEADER;
+        }
+
+        return uxResult;
+    }
+#else /* if ( ipconfigUSE_IPv6 != 0 ) */
+
+/**
+ * @brief Get the size of an IPv4-header, independent of the socket setting.
+ * @param[in] pxSocket: The socket ( ignored ).
+ * @return The size of an IPv4-header.
+ */
+    size_t uxIPHeaderSizeSocket( const FreeRTOS_Socket_t * pxSocket )
```

Comment:
Why not make this a #define like above?

Reply:
`uxIPHeaderSizeSocket` has become a global function, used from other modules as well.

Reply:
Makes sense! Thanks!